File: util/describe_experiment.pl
#!/usr/bin/env perl
#
# Given an experiment configuration file and several Makefile settings, this will
# create a short description

use strict;
use warnings;
use autodie;
use File::Basename;
use File::stat;
use File::Slurp;
use Getopt::Long;

my $USAGE = "Usage: ./$0 [-t TRAINING_SET] [-e] [-d] [-c CV] [-r] config.py \n";

my ( $eval_data, $training_set, $debug, $cv, $rands, $portion ) = ( 0, '', '', 0, '', 0, 1.0 );
GetOptions(
    'training_set|training|t=s' => \$training_set,
    'debug|d'                   => \$debug,
    'cv_runs|cv|c=s'            => \$cv,
    'rands|r'                   => \$rands,
    'train_portion|portion|p=f' => \$portion,
    'eval_data|eval|e'          => \$eval_data,
) or die($USAGE);
die($USAGE) if ( !@ARGV );

# Gather the settings from the command arguments and config files
my ( $data_set, $iters, $training_data, $gadgets, $run_setting, $nn_shape ) = ( '', '', '', '', '', '' );
my $config_data = read_file( $ARGV[0] );

# remove commented-out lines
$config_data =~ s/^\s*#.*$//gm;

# remove classification filter data so that they do not influence reading other settings
my $valid_crit_data = ( $config_data =~ /'validation_weights'\s*:\s*{([^}]*)}/s )[0];
$config_data =~ s/'validation_weights'\s*:\s*{[^}]*}//s;

# data set (devel -- default, eval -- mark)
if ($eval_data){
    $data_set = "\e[1;31mE\e[0m ";
}

# iterations
$iters = ( $config_data =~ /'passes'\s*:\s*([0-9]+)\s*,/ )[0];
if ( $config_data =~ /'pretrain_passes'\s*:\s*([0-9]+)\s*,/ and $1 > 0){
    $iters = $1 . '^'. $iters;
}
if ( $config_data =~ /'use_seq2seq'\s*:\s*True/ ){
    $iters = 'S' . (( $config_data =~ /'seq2seq_pretrain_passes'\s*:\s*([0-9]+)\s*,/ )[0] // 0 ) . ' ' . $iters;
}
if ((( $config_data =~ /'daclassif_pretrain_passes'\s*:\s*([0-9]+)\s*,/ )[0] // 0 ) > 0){
    $iters = 'D' . (( $config_data =~ /'daclassif_pretrain_passes'\s*:\s*([0-9]+)\s*,/ )[0] // 0 ) . ' ' . $iters;
}
$iters .= '/' . ( $config_data =~ /'batch_size'\s*:\s*([0-9]+)\s*,/ )[0];
$iters .= '/' . ( $config_data =~ /'alpha'\s*:\s*([.0-9eE-]+)\s*,/ )[0];
if ( $config_data =~ /'alpha_decay'\s*:\s*([.0-9eE-]+)\s*,/ and $1 > 0){
    $iters .= '^' . $1;
}
$iters =~ s/\/\//\/~\//;
$iters =~ s/\/$/\/~/;

if ($config_data =~ /'validation_size'\s*:\s*([0-9]+)\s*,/ and $1 != 0 ){
    $iters = ( ( $config_data =~ /'min_passes'\s*:\s*([0-9]+)\s*,/ )[0] // 1 ) . '-' . $iters;
    $iters .= ' V' . ( $config_data =~ /'validation_size'\s*:\s*([0-9]+)\s*,/ )[0];
    $iters .= '@' . ( ( $config_data =~ /'validation_freq'\s*:\s*([0-9]+)\s*,/ )[0] // 10);
    $iters .= ' I' . ( ( $config_data =~ /'improve_interval'\s*:\s*([0-9]+)\s*,/ )[0] // 10);
    $iters .= '@' . ( ( $config_data =~ /'top_k'\s*:\s*([0-9]+)\s*,/ )[0] // 5);

    if ($valid_crit_data){
        $valid_crit_data =~ s/[\s\r\n:']//g;
        $valid_crit_data =~ s/([a-z])[a-z]*(?=[^a-z])/$1/g;
        $valid_crit_data =~ s/_//g;
        $valid_crit_data =~ s/,$//;
        $iters .= ' ' . $valid_crit_data;
    }
}

# data style
if ($portion < 1.0){
    $training_set .= '/' . $portion;
}
$training_set .= ' -slotn' if ( $config_data =~ /'delex_slot_names'\s*:\s*True/ );
$training_set .= ' -dlxda' if ( $config_data =~ /'delex_das'\s*:\s*True/ );
$training_set .= ' +lex' if ( $config_data !~ /'delex_slots'\s*:\s*'[^']/ );

my $target_col = 'Q';  # quality is the default
if ($config_data =~ /'target_col'\s*:\s*\[\s*([^\]]*)\]/){
    $target_col = $1;
    $target_col =~ s/['\s,]+/ /g;
    $target_col =~ s/\b([a-z])[a-z]*/$1/g;
    $target_col =~ s/ //g;
    $target_col = uc($target_col);
}
elsif ($config_data =~ /'target_col'\s*:\s*'([^']*)'/){
    $target_col = $1;
    $target_col = uc(substr($target_col, 0, 1));
}
$training_data = $training_set . ' -> ' . $target_col;


# gadgets
$nn_shape .= ' E' . ( ( $config_data =~ /'emb_size'\s*:\s*([0-9]*)/ )[0] // 50 );
$nn_shape .= '-T' . ( ( $config_data =~ /'tanh_layers'\s*:\s*([0-9]*)/ )[0] // 0 );

if ( $config_data =~ /'dropout_keep_prob'\s*:\s*(0\.[0-9]*)/ ){
    $nn_shape .= '-D' . ( $config_data =~ /'dropout_keep_prob'\s*:\s*(0\.[0-9]*)/ )[0];
}
$nn_shape .= ' ' . ( ( $config_data =~ /'cell_type'\s*:\s*'([^']*)'/ )[0] // 'lstm' );
$nn_shape .= ' +bidi'  if ( $config_data =~ /'bidi'\s*:\s*True/ );
$nn_shape .= ' +w2v-s'  if ( $config_data =~ /'word2vec_embs'\s*:\s*'(?!trainable')/ );
$nn_shape .= ' +w2v-t'  if ( $config_data =~ /'word2vec_embs'\s*:\s*'trainable'/ );
$nn_shape .= ' +ce'  if ( $config_data =~ /'char_embs'\s*:\s*True/ );
$nn_shape .= ' +reuse'  if ( $config_data =~ /'reuse_embeddings'\s*:\s*True/ );
$nn_shape .= ' +da'  if ( $config_data =~ /'da_enc'\s*:\s*True/ );
$nn_shape .= ' -ref'  if ( $config_data =~ /'ref_enc'\s*:\s*False/ );
$nn_shape .= ' -hyp'  if ( $config_data =~ /'hyp_enc'\s*:\s*False/ );
$nn_shape .= ' +1/2s'  if ( $config_data =~ /'predict_halves'\s*:\s*True/ );
$nn_shape .= ' +co-t'  if ( $config_data =~ /'predict_coarse'\s*:\s*'train'/ );
$nn_shape .= ' +co-e'  if ( $config_data =~ /'predict_coarse'\s*:\s*'test'/ );
$nn_shape .= ' +ints'  if ( $config_data =~ /'predict_ints'\s*:\s*True/ );
$nn_shape .= ' +adgr'  if ( $config_data =~ /'optimizer_type'\s*:\s*'adagrad'/ );

if ($cv) {
    my @cv_runs = split /\s+/, $cv;
    if (@cv_runs > 1){
        $run_setting .= ' ' . scalar(@cv_runs) . 'CV';
    }
}
if ($debug) {
    $run_setting .= ' DEBUG';
}
if ($rands) {
    $run_setting .= ' RANDS';
}
$run_setting =~ s/^ //;
$run_setting =~ s/ +/,/g;

# Print the output.
print "$training_data $data_set$iters$gadgets$nn_shape ($run_setting)";

File: util/describe_experiment.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function
from __future__ import unicode_literals

from argparse import ArgumentParser
import codecs
import yaml
import re

def main(args):

    iters, training_set, run_setting, nn_shape = '', '', '', ''

    with codecs.open(args.config_file, 'r', 'UTF-8') as fh:
        cfg = yaml.load(fh)

    iters = str(cfg.get('passes', '~'))
    if cfg.get('pretrain_passes'):
        iters = str(cfg['pretrain_passes']) + '^' + iters
    if cfg.get('use_seq2seq'):
        iters = 'S' + str(cfg.get('seq2seq_pretrain_passes', 0)) + ' ' + iters
    if cfg.get('daclassif_pretrain_passes'):
        iters = 'D' + str(cfg.get('daclassif_pretrain_passes')) + ' ' + iters
    iters += '/' + str(cfg.get('batch_size', '~'))
    iters += '/' + (('%.e' % cfg['alpha']).replace('e-0', 'e-') if 'alpha' in cfg else '~')
    if cfg.get('alpha_decay'):
        iters += '^' + str(cfg['alpha_decay'])

    if cfg.get('validation_size'):
        iters = str(cfg.get('min_passes', 1)) + '-' + iters
        iters += ' V' + str(cfg['validation_size'])
        iters += '@' + str(cfg.get('validation_freq', 10))
        iters += ' I' + str(cfg.get('improve_interval', 10))
        iters += '@' + str(cfg.get('top_k', 5))

        if 'validation_weights' in cfg:
            iters += ' ' + ','.join(sorted([
                re.sub('([a-z])[a-z]+(?:_|$)', r'\1', quant.replace('_avg', '')) + str(weight).replace('.0', '')
                for quant, weight in cfg['validation_weights'].iteritems()
            ]))

    training_set = args.training_set
    if args.train_portion < 1.0:
        training_set += '/' + str(args.training_portion)
    training_set += ' -slotn' if cfg.get('delex_slot_names') else ''
    training_set += ' -dlxda' if cfg.get('delex_das') else ''
    training_set += ' +lex' if not cfg.get('delex_slots') else ''

    target_col = 'Q'  # quality is the default
    if 'target_col' in cfg:
        if isinstance(cfg['target_col'], list):
            target_col = ''.join([colname[0].upper() for colname in cfg['target_col']])
        else:
            target_col = cfg['target_col'][0].upper()
    training_set += ' -> ' + target_col

    # gadgets
    nn_shape += ' E' + str(cfg.get('emb_size', 50))
    nn_shape += '-T' + str(cfg.get('tanh_layers', 0))
    if 'dropout_keep_prob' in cfg:
        nn_shape += '-D' + str(cfg['dropout_keep_prob'])

    nn_shape += ' ' + cfg.get('cell_type', 'lstm')
    nn_shape += ' +bidi' if cfg.get('bidi') else ''
    nn_shape += ' +w2v-s' if cfg.get('word2vec_embs') not in [None, 'trainable'] else ''
    nn_shape += ' +w2v-t' if cfg.get('word2vec_embs') == 'trainable' else ''
    nn_shape += ' +ce' if cfg.get('char_embs') else ''
    nn_shape += ' +reuse' if cfg.get('reuse_embeddings') else ''
    nn_shape += ' +da' if cfg.get('da_enc') else ''
    nn_shape += ' -ref' if not cfg.get('ref_enc', True) else ''
    nn_shape += ' -hyp' if not cfg.get('hyp_enc', True) else ''
    nn_shape += ' +1/2s' if cfg.get('predict_halves') else ''
    nn_shape += ' +co-t' if cfg.get('predict_coarse') == 'train' else ''
    nn_shape += ' +co-e' if cfg.get('predict_coarse') == 'test' else ''
    nn_shape += ' +ints' if cfg.get('predict_ints') else ''
    nn_shape += ' +adgr' if cfg.get('optimizer_type') == 'adagrad' else ''
    nn_shape += ' +sqrl' if cfg.get('rank_loss_type') == 'squared' else ''

    if args.cv_runs:
        num_runs = len(args.cv_runs.split())
        run_setting += ' ' + str(num_runs) + 'CV'
    if args.debug:
        run_setting += ' DEBUG'
    if args.rands:
        run_setting += ' RANDS'

    run_setting = run_setting.strip()
    run_setting = run_setting.replace(' ', ',')
    run_setting = ' (' + run_setting + ')' if run_setting else ''

    print(training_set + ' ' + iters + nn_shape + run_setting, end='')


if __name__ == '__main__':

    ap = ArgumentParser()
    ap.add_argument('-t', '--training-set', '--training', type=str, help='Training set name')
    ap.add_argument('-d', '--debug', action='store_true', help='Are we running with debug prints?')
    ap.add_argument('-c', '--cv-runs', '--cv', type=str, help='Number of CV runs used')
    ap.add_argument('-r', '--rands', action='store_true', help='Are we using more random inits?')
    ap.add_argument('-p', '--train-portion', '--portion', type=float, help='Training data portion used', default=1.0)
    ap.add_argument('-e', '--eval-data', '--eval', action='store_true', help='Using evaluation data')
    ap.add_argument('config_file', type=str, help='Experiment YAML config file')

    args = ap.parse_args()
    main(args)
